src/shared/particle_dynamics/solid_dynamics/thin_structure_dynamics.cpp

```
@@ -139,12 +139,33 @@ void ShellStressRelaxationFirstHalf::initialization(size_t index_i, Real dt)
+          /// Differential geometry and the geometrically non-linear Reissner-Mindlin shell model
+          /// @WARN Algorithm is not guaranteed to converge, see discussion in Sec. 5.4.1
+          ///       even more so considering we do not reuse analytical derivatives.
 
-        /** correct out-plane numerical damping. */
-        Matd cauchy_stress = elastic_solid_.StressCauchy(current_local_almansi_strain, F_gaussian_point, index_i) + current_transformation_matrix * transformation_matrix_[index_i].transpose() * F_gaussian_point *
-                                                                                                                        elastic_solid_.NumericalDampingRightCauchy(F_gaussian_point, dF_gaussian_point_dt, numerical_damping_scaling_[index_i], index_i) * F_gaussian_point.transpose() * transformation_matrix_[index_i] * current_transformation_matrix.transpose() / F_gaussian_point.determinant();
+            double E = E0_; // Take the default Young's modulus of the material as the initial derivative.
+            int it = 0;
+            int max_it = 20;
+            for (double s_next = 1.0; abs(s_next) > 1e-9 && it < max_it; ++it) // @WARN Hardcoded tolerance and iteration limit.
+            {
+                double e_prev = current_local_almansi_strain(2, 2);
+                double s_prev = cauchy_stress(2, 2);
+                double e_next = e_prev - s_prev / E;
+                current_local_almansi_strain(2, 2) = e_next;
+                cauchy_stress = elastic_solid_.StressCauchy(current_local_almansi_strain, F_gaussian_point, index_i);
+                s_next = cauchy_stress(2, 2);
+                E = (s_next - s_prev) / (e_next - e_prev);
+            }
+            if (cauchy_stress.allFinite() == false || it == max_it)
+                throw std::runtime_error("Algorithm enforcing plane stress condition for shell failed.");
```

Discussion: https://github.com/Virtonomy/SPHinXsys/pull/123#discussion_r1847888115

---

BTW I will not change it to an error message because it is an exceptional hard failure situation and it must be debugged rather than allowed to continue.

The convergence criterion is up for debate though